rclcpp_generic/include/rclcpp_generic/generic_subscription.hpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_GENERIC__GENERIC_SUBSCRIPTION_HPP_
+#define RCLCPP_GENERIC__GENERIC_SUBSCRIPTION_HPP_
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/macros.hpp"
+#include "rclcpp/subscription_base.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/qos.hpp"
+#include "rclcpp/callback_group.hpp"
+#include "rclcpp/serialized_message.hpp"
+#include "rcpputils/shared_library.hpp"
+
+namespace rclcpp_generic
+{
+
+/**
+ * This class is an implementation of an rclcpp::SubscriptionBase for serialized messages whose topic
+ * is not known at compile time (hence templating does not work).
+ *
+ * It does not support intra-process handling.
+ */
+class GenericSubscription : public rclcpp::SubscriptionBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(GenericSubscription)
+
+  /**
+   * Factory method.
+   *
+   * The returned pointer will never be empty, but this function can throw various exceptions, for
+   * instance when the message's package can not be found on the AMENT_PREFIX_PATH.
+   *
+   * \param topics_interface NodeTopicsInterface pointer used in parts of the setup.
+   * \param topic_name Topic name
+   * \param topic_type Topic type
+   * \param qos QoS settings
+   * \param callback Callback for new messages of serialized form
+   * \param group Callback group
+   */
+  static std::shared_ptr<GenericSubscription> create(
```

Comment:
Is it better to have a static member function or a free function?

Reply:
Static member seems nice given that it's tightly tied with the class and this is C++. Free function would definitely be the C approach :)